### ISSUE TITLE
Question about code flow

### ISSUE THREAD

**Issue by MonsKim999** [jialinwu17/generate_captions_for_vqa#2]
Hi jialinwu.
Thank you for opening your great code before asking.

I understood your paper.
first: pretrained vqa and caption model.
second: perform the multi-task learning for the question-related caption model and vqa
third: at the testing, use the newly generated caption as input C and inference the A.
Thanks for letting me know if I'm misunderstanding.

My question is below:

1. In the finetune_vqa_model.py, there is an argument as the name of 'task'
and this is set by default 'VQA'. but I don't understand why the output of the model has three.
what's the meaning of loss,loss1 and loss2 ?

2. If I want to learn the task of 'caption' , Do I have to change the argument of task as 'caption'?
If It is right, like common sense, I think It is right to learn jointly like mulit-tasking. but I cannot find where the code set like this.

3. What is the difference between captions and rcaptions?

4. What is the role of features_masks?

best regards.

**Comment by jialinwu17** [comment]
Q1: loss is using Object+Question+Caption, loss2 is using Question+Caption, and  loss3 is using Question+Object. During experiments, I found just using loss without loss1 and loss2 provides the best results.
Q2: To learn to generate captions, I used the code train_caption_model.py. 
Q3: Caption is padding on the left and rcaption is padding on the right side.  For example, if a caption is like ``I want to play'' Captions  gives ``<pad><pad><pad>I want to play'' and rcaption gives ``I want to play<pad><pad><pad>''
Caption is used as features so we want the last tokens to be content, but rcaption is used as supervision in the captioning task so we want the first tokens to be content.
Q4: feature masks are used to indicate if the objects appear. I don't think it is useful in the current repo but it is useful when we have  various number of objects say from 10-100(currently we have a fixed number of 36)
